### ISSUE TITLE
pipeenv can't find public package when using private gitlab repo

### ISSUE THREAD

**Issue by Kroppeb** [pypa/pipenv#6092]
We have a private GitLab package repo with request forwarding.

Having the following in a pipfile works:
```
[[source]]
# private source

[dev-packages]
opencv-python = "4.8"
numpy = "*"
```

but this fails:

```
[[source]]
# private source

[dev-packages]
opencv-python = "4.8"
```

Because it can't find the `numpy` package (dependency of opencv)

Here is the verbose logs:

<details>
<summary> Good one </summary>

```
Locking [packages] dependencies...
Locking [dev-packages] dependencies...
Building requirements...
Resolving dependencies...
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting()
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
'), None)
INFO:pipenv.patched.pip._internal.operations.prepare:Collecting numpy (from -r
C:\Users\Robbe\AppData\Local\Temp\pipenv-wuskap5a-requirements\pipenv-5v58bkyu-constraints.txt (line 3))
INFO:pipenv.patched.pip._internal.network.download:Using cached numpy-1.26.4-cp310-cp310-win_amd64.whl.metadata (61 kB)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('openc
v-python==4.8.*'), None)
INFO:pipenv.patched.pip._internal.operations.prepare:Collecting opencv-python==4.8.* (from -r
C:\Users\Robbe\AppData\Local\Temp\pipenv-wuskap5a-requirements\pipenv-5v58bkyu-constraints.txt (line 2))
INFO:pipenv.patched.pip._internal.network.download:Using cached opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl.metadata
(20 kB)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(0)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.21.2; python_version >= "3.10"'),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.17.0; python_version >= "3.7"'),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.17.3; python_version >= "3.8"'),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.19.3; python_version >= "3.9"'),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.pinning(LinkCandidate('https://files.pythonhos
ted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72f/opencv_python-4.8.1.78-cp37-abi3-win
_amd64.whl (from https://pypi.org/simple/opencv-python/) (requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending_round(0, state)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(1)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.pinning(LinkCandidate('https://files.pythonhos
ted.org/packages/19/77/538f202862b9183f54108557bfda67e17603fc560c384559e769321c9d92/numpy-1.26.4-cp310-cp310-win_amd64.w
hl (from https://pypi.org/simple/numpy/) (requires-python:>=3.9)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending_round(1, state)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(2)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending(State(mapping=OrderedDict([('opencv-pyt
hon',
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), ('numpy',
LinkCandidate('https://files.pythonhosted.org/packages/19/77/538f202862b9183f54108557bfda67e17603fc560c384559e769321c9d9
2/numpy-1.26.4-cp310-cp310-win_amd64.whl (from https://pypi.org/simple/numpy/) (requires-python:>=3.9)'))]),
criteria={'numpy': Criterion((SpecifierRequirement('numpy'), via=None), (SpecifierRequirement('numpy>=1.21.2;
python_version >= "3.10"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac390
4e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), (SpecifierRequirement('numpy>=1.17.0; python_version >= "3.7"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac390
4e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), (SpecifierRequirement('numpy>=1.17.3; python_version >= "3.8"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac390
4e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), (SpecifierRequirement('numpy>=1.19.3; python_version >= "3.9"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac390
4e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))), 'opencv-python': Criterion((SpecifierRequirement('opencv-python==4.8.*'), via=None))},
backtrack_causes=[]))
INFO:pipenv.patched.pip._internal.network.download:Using cached opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (38.1 MB)
INFO:pipenv.patched.pip._internal.network.download:Using cached numpy-1.26.4-cp310-cp310-win_amd64.whl (15.8 MB)
[=   ] Locking...
Success!
Updated Pipfile.lock (bf0fa0e85a8ed1123a35bf320caf73145295cea203de6d43bc416a2e72088a63)!
Installing dependencies from Pipfile.lock (088a63)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
```

</details>

<details>
<summary> Bad one </summary>

```
Locking [packages] dependencies...
Locking [dev-packages] dependencies...
Building requirements...
Resolving dependencies...
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting()
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('openc
v-python==4.8.*'), None)
INFO:pipenv.patched.pip._internal.operations.prepare:Collecting opencv-python==4.8.* (from -r
C:\Users\Robbe\AppData\Local\Temp\pipenv-dl6b59le-requirements\pipenv-3_3j_u0y-constraints.txt (line 2))
INFO:pipenv.patched.pip._internal.network.download:Using cached opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl.metadata
(20 kB)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(0)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.21.2; python_version >= "3.10"'),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
WARNING:pipenv.patched.pip._vendor.urllib3.connectionpool:Retrying (Retry(total=4, connect=None, read=None,
redirect=None, status=None)) after connection broken by
'NewConnectionError('<pipenv.patched.pip._vendor.urllib3.connection.HTTPSConnection object at 0x0000017B98F2A950>:
Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /numpy/
WARNING:pipenv.patched.pip._vendor.urllib3.connectionpool:Retrying (Retry(total=3, connect=None, read=None,
redirect=None, status=None)) after connection broken by
'NewConnectionError('<pipenv.patched.pip._vendor.urllib3.connection.HTTPSConnection object at 0x0000017B98F2BE80>:
Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /numpy/
WARNING:pipenv.patched.pip._vendor.urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None,
redirect=None, status=None)) after connection broken by
'NewConnectionError('<pipenv.patched.pip._vendor.urllib3.connection.HTTPSConnection object at 0x0000017B98F2BDC0>:
Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /numpy/
WARNING:pipenv.patched.pip._vendor.urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None,
redirect=None, status=None)) after connection broken by
'NewConnectionError('<pipenv.patched.pip._vendor.urllib3.connection.HTTPSConnection object at 0x0000017B9A751180>:
Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /numpy/
WARNING:pipenv.patched.pip._vendor.urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None,
redirect=None, status=None)) after connection broken by
'NewConnectionError('<pipenv.patched.pip._vendor.urllib3.connection.HTTPSConnection object at 0x0000017B9A752F80>:
Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /numpy/
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.rejecting_candidate(Criterion((SpecifierRequir
ement('numpy>=1.21.2; python_version >= "3.10"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac390
4e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))),
LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac3904e72
f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.network.download:Using cached opencv_python-4.8.0.76-cp37-abi3-win_amd64.whl.metadata
(20 kB)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.21.2; python_version >= "3.10"'),
LinkCandidate('https://files.pythonhosted.org/packages/fb/c4/f574ba6f04e6d7bf8c38d23e7a52389566dd7631fee0bcdd79ea07ef2db
f/opencv_python-4.8.0.76-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.rejecting_candidate(Criterion((SpecifierRequir
ement('numpy>=1.21.2; python_version >= "3.10"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/fb/c4/f574ba6f04e6d7bf8c38d23e7a52389566dd7631fee0bcdd79ea07e
f2dbf/opencv_python-4.8.0.76-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))),
LinkCandidate('https://files.pythonhosted.org/packages/fb/c4/f574ba6f04e6d7bf8c38d23e7a52389566dd7631fee0bcdd79ea07ef2db
f/opencv_python-4.8.0.76-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.network.download:Using cached opencv_python-4.8.0.74-cp37-abi3-win_amd64.whl.metadata
(19 kB)
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('numpy
>=1.21.2; python_version >= "3.10"'),
LinkCandidate('https://files.pythonhosted.org/packages/a6/08/583e1d8d589e80a702815f58daa317bd469a32c7dc339394e2a0dc14318
b/opencv_python-4.8.0.74-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.rejecting_candidate(Criterion((SpecifierRequir
ement('numpy>=1.21.2; python_version >= "3.10"'),
via=LinkCandidate('https://files.pythonhosted.org/packages/a6/08/583e1d8d589e80a702815f58daa317bd469a32c7dc339394e2a0dc1
4318b/opencv_python-4.8.0.74-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))),
LinkCandidate('https://files.pythonhosted.org/packages/a6/08/583e1d8d589e80a702815f58daa317bd469a32c7dc339394e2a0dc14318
b/opencv_python-4.8.0.74-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))
CRITICAL:pipenv.patched.pip._internal.resolution.resolvelib.factory:Cannot install -r
C:\Users\Robbe\AppData\Local\Temp\pipenv-dl6b59le-requirements\pipenv-3_3j_u0y-constraints.txt (line 2) because these
package versions have conflicting dependencies.
INFO:pipenv.patched.pip._internal.resolution.resolvelib.factory:
The conflict is caused by:
    opencv-python 4.8.1.78 depends on numpy>=1.21.2; python_version >= "3.10"
    opencv-python 4.8.0.76 depends on numpy>=1.21.2; python_version >= "3.10"
    opencv-python 4.8.0.74 depends on numpy>=1.21.2; python_version >= "3.10"
To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict
Traceback (most recent call last):
  File
"C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\patched\pip\_internal\resolution\resolv
elib\resolver.py", line 95, in resolve
    result = self._result = resolver.resolve(
  File
"C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\patched\pip\_vendor\resolvelib\resolver
s.py", line 546, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File
"C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\patched\pip\_vendor\resolvelib\resolver
s.py", line 439, in resolve
    raise ResolutionImpossible(self.state.backtrack_causes)
pipenv.patched.pip._vendor.resolvelib.resolvers.ResolutionImpossible:
[RequirementInformation(requirement=SpecifierRequirement('numpy>=1.21.2; python_version >= "3.10"'),
parent=LinkCandidate('https://files.pythonhosted.org/packages/38/d2/3e8c13ffc37ca5ebc6f382b242b44acb43eb489042e1728407ac
3904e72f/opencv_python-4.8.1.78-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), RequirementInformation(requirement=SpecifierRequirement('numpy>=1.21.2; python_version >=
"3.10"'),
parent=LinkCandidate('https://files.pythonhosted.org/packages/fb/c4/f574ba6f04e6d7bf8c38d23e7a52389566dd7631fee0bcdd79ea
07ef2dbf/opencv_python-4.8.0.76-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)')), RequirementInformation(requirement=SpecifierRequirement('numpy>=1.21.2; python_version >=
"3.10"'),
parent=LinkCandidate('https://files.pythonhosted.org/packages/a6/08/583e1d8d589e80a702815f58daa317bd469a32c7dc339394e2a0
dc14318b/opencv_python-4.8.0.74-cp37-abi3-win_amd64.whl (from https://pypi.org/simple/opencv-python/)
(requires-python:>=3.6)'))]
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 446, in
resolve
    results = resolver.resolve(constraints, check_supported_wheels=False)
  File
"C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\patched\pip\_internal\resolution\resolv
elib\resolver.py", line 104, in resolve
    raise error from e
pipenv.patched.pip._internal.exceptions.DistributionNotFound: ResolutionImpossible: for help visit
https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\resolver.py", line 675, in
<module>
    main()
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\resolver.py", line 661, in main
    _main(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\resolver.py", line 645, in _main
    resolve_packages(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\resolver.py", line 612, in
resolve_packages
    results, resolver = resolve(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\resolver.py", line 592, in
resolve
    return resolve_deps(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 918, in
resolve_deps
    results, hashes, internal_resolver = actually_resolve_deps(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 691, in
actually_resolve_deps
    resolver.resolve()
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 448, in
resolve
    raise ResolutionFailure(message=str(e))
pipenv.exceptions.ResolutionFailure: [31m[1mERROR[0m: [33mResolutionImpossible: for help visit
https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts[0m
Locking Failed!
[==  ] Locking...
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Scripts\pipenv.exe\__main__.py", line 7, in <module>
    # when invoked as python -m pip <command>
                 ^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\cli\options.py", line 58, in main
    return super().main(*args, **kwargs, windows_expand_args=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\decorators.py", line 92, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\vendor\click\core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\cli\command.py", line 209, in install
    do_install(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\routines\install.py", line 164, in do_install
    do_init(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\routines\install.py", line 648, in do_init
    do_lock(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\routines\lock.py", line 65, in do_lock
    venv_resolve_deps(
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 859, in venv_resolve_deps
    c = resolve(cmd, st, project=project)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Robbe\AppData\Local\Programs\Python\Python312\Lib\site-packages\pipenv\utils\resolver.py", line 728, in resolve
    raise RuntimeError("Failed to lock Pipfile.lock!")
RuntimeError: Failed to lock Pipfile.lock!
```

</details>

**Comment by matteius** [comment]
This is known -- the way index package restrictions was implemented (to prevent package confusion attacks) always defaults unspecified package indexes to the first index in the Pipfile.

**Comment by Kroppeb** [comment]
To clarify, the private source is the only source in the pipfile.

**Comment by matteius** [comment]
Well are you mirroring packages from pypi in your private source?

**Comment by Kroppeb** [comment]
Yeah, numpy and opencv both install correctly when explicitly mentioned as a dependency. But when numpy isn't mentioned and instead downloaded as a dependency of a dependency, something seems to go wrong.

**Comment by matteius** [comment]
Other gitlab dependency problems have recently been resolved with `2024.3.0` -- closing this for now, and we can revisit if its still an issue for you @Kroppeb
